### PR TITLE
KAFKA-6753: Updating the OfflinePartitions count only when necessary

### PR DESCRIPTION
**Pull request by gitlw**
Instead of updating the offline partitions count blindly after processing any event in the controller, 
this patch maintains the offline partitions count in a variable inside the PartitionStateMachine,
and only updates it when necessary.
This will be the first step in optimizing the KafkaController#updateMetrics function.

Testing done:
1. Added new test methods for the new way of updating offline partitions count.
2. The new logic has been used inside LinkedIn in production for a few months without any problem.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by gitlw** [comment]
retest this please

**Comment by gitlw** [comment]
@becketqin Can you please take a look at this PR when you have time? Thanks a lot!

File: core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala [hunk 1]
```diff
@@ -17,13 +17,15 @@
 package kafka.controller
 
 import kafka.api.LeaderAndIsr
+import kafka.common.TopicAndPartition
```
Comment (omkreddy) [inline core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala]: unused imports?

File: core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala [hunk 2]
```diff
@@ -36,6 +38,7 @@ import scala.collection.mutable
 class PartitionStateMachineTest extends JUnitSuite {
   private var controllerContext: ControllerContext = null
   private var mockZkClient: KafkaZkClient = null
+  private var mockZkUtils: ZkUtils = null
```
Comment (omkreddy) [inline core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala]: unused variable

File: core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala [hunk 3]
```diff
@@ -312,4 +316,75 @@ class PartitionStateMachineTest extends JUnitSuite {
...
+
+  @Test
+  def testNoOfflinePartitionsChangeForTopicsBeingDeleted() = {
+    val partitionIds = Seq(0, 1, 2, 3)
+    val topic = "test"
+    val partitions = partitionIds.map(new TopicPartition("test", _))
+
+    EasyMock.expect(mockTopicDeletionManager.isTopicQueuedUpForDeletion(topic)).andReturn(true)
+    EasyMock.expectLastCall().anyTimes()
+
```
Comment (omkreddy) [inline core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala]: EasyMock.replay() statement is missing

File: core/src/main/scala/kafka/controller/PartitionStateMachine.scala [hunk 4]
```diff
@@ -125,6 +128,21 @@ class PartitionStateMachine(config: KafkaConfig,
...
   }
 
+  private def changeStateTo(partition: TopicPartition, currentState: PartitionState, targetState: PartitionState): Unit = {
+    partitionState.put(partition, targetState)
+    updateControllerMetrics(partition, currentState, targetState)
+  }
+
+  private def updateControllerMetrics(partition: TopicPartition, currentState: PartitionState, targetState: PartitionState) : Unit = {
+    if (!topicDeletionManager.isTopicQueuedUpForDeletion(partition.topic)) {
+      if (currentState != OfflinePartition && targetState == OfflinePartition) {
```
Comment (omkreddy) [inline core/src/main/scala/kafka/controller/PartitionStateMachine.scala]: looks like we are not checking "!controllerContext.liveOrShuttingDownBrokerIds.contains()" ,  which was there in old code. Is that check not required now?

**Comment by harshach** [comment]
@gitlw thanks for the patch. But your PR shows past commits and huge file update list. Can you make sure you rebased against master along with your changes on top and re-open the pr.

File: core/src/main/scala/kafka/controller/PartitionStateMachine.scala (hunk 4 above)
Comment (gitlw) [reply core/src/main/scala/kafka/controller/PartitionStateMachine.scala]: @omkreddy Thanks for bringing this up. I discussed internally with someone in our team, and we think the check is not required now. The old definition of an offline partition is one whose leader is not alive, and the new definition is one whose state in the PartitionStateMachine is Offline. Comparing these definitions gives us 4 combinations
1. leader alive and state in not offline => both definitions count the partition as not offline
2. leader alive and state is offline
3. no alive leader and state is not offline
4. no alive leader and state is offline => both definitions count the partition as offline

In case 1 and 4 these definitions give the same result.
For case 2: by checking the code, it seems as long as there is an alive leader during events like TopicChange, BrokerFailure or ControlledShutdown, by the end of processing such events, the controller will always change the state of the partition to Online. So it seems under normal operations by the end of processing any event, case 2 will never happen. 
I also checked the failure scenario when the request to update leadership info through Zookeeper fails, then the controllerContext.partitionLeadershipInfo is not updated, and the state in PartitionStateMachine is not updated. The end result is that both definitions will use the result before the zk update.

For case 3, not being in offline state gives us 3 other possibilities, i.e. NewPartition, OnlinePartition, and NonExistentPartition. 
- NewPartition is only being used as an intermediate state, and no partition should be in this state after processing any event.
- For OnlinePartition, it should not happen when there is no alive leader.
- For NonExistentPartition, it only happens at the last step of topic deletion, at which point the leader should also be deleted. So this combination is also impossible.

In summary, the new definition should not give different results compared with the new definition by the end of processing any event. 
Please let me know if anything is not clear. Thanks!

File: core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala (hunk 1 above)
Comment (gitlw) [reply core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala]: Good catch. It's fixed now.

File: core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala (hunk 2 above)
Comment (gitlw) [reply core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala]: Fixed now. Thanks.

**Comment by gitlw** [comment]
@omkreddy Thanks for your comments. I've addressed them. Please take another look when you have time. Thank you!

**Comment by ijuma** [comment]
@junrao Seems like this is important to merge given the issue described in the JIRA.

File: core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala [hunk 5]
```diff
@@ -19,7 +19,7 @@ package kafka.controller
 import kafka.api.LeaderAndIsr
 import kafka.log.LogConfig
 import kafka.server.KafkaConfig
-import kafka.utils.TestUtils
+import kafka.utils.{TestUtils, ZkUtils}
```
Comment (omkreddy) [inline core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala]: unused import

File: core/src/main/scala/kafka/controller/PartitionStateMachine.scala (hunk 4 above)
Comment (omkreddy) [reply core/src/main/scala/kafka/controller/PartitionStateMachine.scala]: Thanks for the explanation. Sounds good to me.

File: core/src/main/scala/kafka/controller/PartitionStateMachine.scala [hunk 6]
```diff
@@ -52,6 +52,8 @@ class PartitionStateMachine(config: KafkaConfig,
 
   this.logIdent = s"[PartitionStateMachine controllerId=$controllerId] "
 
+  var offlinePartitionCount = 0
```
Comment (junrao) [inline core/src/main/scala/kafka/controller/PartitionStateMachine.scala]: We probably want to make this volatile. The reader of this value is typically the metrics reporter, which will be in a different thread from the controller thread.

File: core/src/main/scala/kafka/controller/PartitionStateMachine.scala [hunk 7]
```diff
@@ -125,6 +128,21 @@ class PartitionStateMachine(config: KafkaConfig,
...
+    partitionState.put(partition, targetState)
+    updateControllerMetrics(partition, currentState, targetState)
+  }
+
+  private def updateControllerMetrics(partition: TopicPartition, currentState: PartitionState, targetState: PartitionState) : Unit = {
+    if (!topicDeletionManager.isTopicQueuedUpForDeletion(partition.topic)) {
+      if (currentState != OfflinePartition && targetState == OfflinePartition) {
+        offlinePartitionCount = offlinePartitionCount + 1
+      } else if (currentState == OfflinePartition && targetState != OfflinePartition) {
+        offlinePartitionCount = offlinePartitionCount - 1
```
Comment (junrao) [inline core/src/main/scala/kafka/controller/PartitionStateMachine.scala]: A couple of points on this approach. (1) Currently, if a broker goes offline, the controller calls onReplicasBecomeOffline() which first moves a partition to offline state and then tries to bring the partition to online immediately after. If a leader can be elected, the partition will be transitioned to the online state. With this PR, one potential issue is that there is a very short window in which the offline count could go up and then go down. This can give false alarms in the alerting system. (2) It seems that if a topic is deleted, we don't necessarily go through the partition state transition. We need to update the offline partition count accordingly.

**Review by junrao — COMMENTED** [review]
@gitlw : Thanks for the patch. Left a couple of comments below.

File: core/src/main/scala/kafka/controller/PartitionStateMachine.scala (hunk 6 above)
Comment (gitlw) [reply core/src/main/scala/kafka/controller/PartitionStateMachine.scala]: Thank for your review, Jun. 

The metrics reporter should be reading directly from the KafkaController.offlinePartitionCount variable, which is already volatile.
This offlinePartitionCount inside PartitionStateMachine should be only used by the controller thread, so I feel it's ok to leave it as non-volatile. What do you think?

File: core/src/main/scala/kafka/controller/PartitionStateMachine.scala (hunk 7 above)
Comment (gitlw) [reply core/src/main/scala/kafka/controller/PartitionStateMachine.scala]: (1) You are right that the PartitionStateMachine.offlinePartitionCount can go up and down for some events, e.g. a broker going offline, but the metrics reporter is NOT using this variable directly.
Instead the metrics reporter uses the KafkaController.offlinePartitionCount, which is only updated at the end of processing controller events, so the intermediate value changes won't be exposed to the metrics. 

(2) Currently once a topic is enqueued for deletion in the TopicDeletionManager.topicsToBeDeleted, we don't care about the status of their partitions, and won't count them as offline partitions. I assume we still want to keep the same behavior.
(2.1) With this PR, if partition is online before deletion (offlinePartitionCount counted as 0), then starting the deletion of its topic and maintain the offlinePartitionCount being 0 is the correct behavior.  
(2.2) However if a partition is offline (its offlinePartitionCount counted as 1) when the topic deletion started, then maintaining the value being 1 does not make sense anymore, since we want to stop counting the deleted partition as being offline. This suggests when enqueuing a topic for deletion, we should get the current offline partitions for these topics, and reverse their count in the PartitionStateMachine.offlinePartitionCount variable. 
Does that sound right to you? Thanks!

File: core/src/main/scala/kafka/controller/PartitionStateMachine.scala (hunk 6 above)
Comment (junrao) [reply core/src/main/scala/kafka/controller/PartitionStateMachine.scala]: Thanks for the explanation. Makes sense.

File: core/src/main/scala/kafka/controller/PartitionStateMachine.scala (hunk 7 above)
Comment (junrao) [reply core/src/main/scala/kafka/controller/PartitionStateMachine.scala]: (1) Makes sense. (2) Yes, 2.2 is the case that I was concerned about. Your suggested fix makes sense.

File: core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala [hunk 8]
```diff
@@ -312,4 +313,143 @@ class PartitionStateMachineTest extends JUnitSuite {
     assertEquals(OfflinePartition, partitionState(partition))
   }
 
+  def prepareMockToElectLeaderForPartitions(partitions: Seq[TopicPartition]): Unit = {
```
Comment (junrao) [inline core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala]: Could this method be private?

File: core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala [hunk 9]
```diff
@@ -312,4 +313,143 @@ class PartitionStateMachineTest extends JUnitSuite {
...
+      val getDataResponses = partitions.map {p => GetDataResponse(Code.OK, null, Some(p),
+        TopicPartitionStateZNode.encode(leaderIsrAndControllerEpoch), stat, ResponseMetadata(0, 0))}
+      EasyMock.expect(mockZkClient.getTopicPartitionStatesRaw(partitions))
+        .andReturn(getDataResponses)
+    }
+    prepareMockToGetTopicPartitionsStatesRaw()
+
+    def prepareMockToGetLogConfigs(): Unit = {
+      val topicsForPartitionsWithNoLiveInSyncReplicas = Seq()
+      //val topicConfigs = topics.map{topic => (topic, LogConfig())}.toMap
```
Comment (junrao) [inline core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala]: Is the commented out code still needed? Ditto below.

File: core/src/main/scala/kafka/controller/TopicDeletionManager.scala [hunk 10]
```diff
@@ -102,8 +102,12 @@ class TopicDeletionManager(controller: KafkaController,
    */
   def enqueueTopicsForDeletion(topics: Set[String]) {
     if (isDeleteTopicEnabled) {
-      topicsToBeDeleted ++= topics
       partitionsToBeDeleted ++= topics.flatMap(controllerContext.partitionsForTopic)
+      // move the respective partitions to OfflinePartition and NonExistentPartition state, so that the
+      // offlinePartitionCount is properly updated
+      controller.partitionStateMachine.handleStateChanges(partitionsToBeDeleted.toSeq, OfflinePartition)
+      controller.partitionStateMachine.handleStateChanges(partitionsToBeDeleted.toSeq, NonExistentPartition)
```
Comment (junrao) [inline core/src/main/scala/kafka/controller/TopicDeletionManager.scala]: Hmm, moving the code here fixes the offlinePartitionCount issue. However, it introduces another issue. If the topic is in pending deletion state because of an going partition reassignment, the partition can't be moved to online state once it's in NonExistentPartition state. This may prevent the reassignment from completing.

**Review by junrao — COMMENTED** [review]
@gitlw : Thanks for the updated patch. A few more comments.

File: core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala (hunk 8 above)
Comment (gitlw) [reply core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala]: Fixed.

File: core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala (hunk 9 above)
Comment (gitlw) [reply core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala]: Removed these lines.

File: core/src/main/scala/kafka/controller/TopicDeletionManager.scala (hunk 10 above)
Comment (gitlw) [reply core/src/main/scala/kafka/controller/TopicDeletionManager.scala]: Good catch!
To address this issue, the latest commit in this PR has the following changes
- Removed the TopicDeletionManager#partitionsToBeDeleted variable since it is always updated together with the topicsToBeDeleted variable, and therefore checking whether a partition is to be deleted can be done through the topicsToBeDelete variable.
- Added a new variable partitionsBeingDeleted, which includes partitions that are going through actual deletions. More partitions are added to it in the TopicDeletionManager#onTopicDeletion method, where the deletion for a partition has started. Once a partition is added to the partitionsBeingDeleted variable, we are sure partition reassignment can NO LONGER happen to it.
- The PartitionStateMachine#updateControllerMetrics is updated to check if a partition is BEING deleted. This means if reassignment is going on for a given partition while the corresponding topic is enqueued for deletion, we will still monitor its state change until the reassignment finishes and the deletion starts.
Let me know what you think, and thanks a lot for your review!

File: core/src/main/scala/kafka/controller/TopicDeletionManager.scala [hunk 11]
```diff
@@ -62,13 +62,12 @@ class TopicDeletionManager(controller: KafkaController,
   val controllerContext = controller.controllerContext
   val isDeleteTopicEnabled = controller.config.deleteTopicEnable
   val topicsToBeDeleted = mutable.Set.empty[String]
-  val partitionsToBeDeleted = mutable.Set.empty[TopicPartition]
+  val partitionsBeingDeleted = mutable.Set.empty[TopicPartition]
```
Comment (junrao) [inline core/src/main/scala/kafka/controller/TopicDeletionManager.scala]: Could we add a comment on the meaning of partitionsBeingDeleted?

File: core/src/main/scala/kafka/controller/TopicDeletionManager.scala [hunk 12]
```diff
@@ -254,6 +248,17 @@ class TopicDeletionManager(controller: KafkaController,
...
+    /** Once a partition is added to the partitionsBeingDeleted variable,
+      * we stop monitoring its state changes in the offlinePartitionCount metric.
+      * Therefore if some partitions are already in OfflinePartition state,
+      * and recorded in the offlinePartitionCount state, we need to properly
+      * decrement their count in offlinePartitionCount before adding them to
+      * the partitionsBeingDeleted set.
+      */
+    controller.partitionStateMachine.handleStateChanges(unseenPartitionsForDeletion.toSeq, OfflinePartition)
+    controller.partitionStateMachine.handleStateChanges(unseenPartitionsForDeletion.toSeq, NonExistentPartition)
+    partitionsBeingDeleted ++= unseenPartitionsForDeletion
```
Comment (junrao) [inline core/src/main/scala/kafka/controller/TopicDeletionManager.scala]: This approach does seem to work overall. A couple of more things. (1) Since we always add/remove all partitions in a topic together, could we just track this at the topic level? (2) In ControlledShutdown.doControlledShutdown(), we call partitionStateMachine.handleStateChanges() on partitions w/o checking whether the topics are being deleted. This seems to be an existing issue. However, with this patch, it will fail the partition state transition once those partitions are moved to NonExist state.

File: core/src/main/scala/kafka/controller/TopicDeletionManager.scala [hunk 13]
```diff
@@ -283,9 +288,9 @@ class TopicDeletionManager(controller: KafkaController,
       val successfullyDeletedReplicas = controller.replicaStateMachine.replicasInState(topic, ReplicaDeletionSuccessful)
       val replicasForDeletionRetry = aliveReplicasForTopic -- successfullyDeletedReplicas
       // move dead replicas directly to failed state
-      controller.replicaStateMachine.handleStateChanges(deadReplicasForTopic.toSeq, ReplicaDeletionIneligible)
+      controller.replicaStateMachine.handleStateChanges(deadReplicasForTopic.toSeq, ReplicaDeletionIneligible, new Callbacks())
```
Comment (junrao) [inline core/src/main/scala/kafka/controller/TopicDeletionManager.scala:310]: This change doesn't seem to be needed? Ditto below.

File: core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala [hunk 14]
```diff
@@ -312,4 +313,147 @@ class PartitionStateMachineTest extends JUnitSuite {
...
+    partitionStateMachine.handleStateChanges(partitions, NewPartition)
+    partitionStateMachine.handleStateChanges(partitions, OfflinePartition)
+    assertEquals(s"There should be ${partitions.size} offline partition(s)", partitions.size, partitionStateMachine.offlinePartitionCount)
+
+    partitionStateMachine.handleStateChanges(partitions, OnlinePartition, Some(OfflinePartitionLeaderElectionStrategy))
+    assertEquals(s"There should be no offline partition(s)", 0, partitionStateMachine.offlinePartitionCount)
+  }
+
+  /**
+    * This method tests if topic deletion is disabled, then changing partitions' state to OfflinePartition makes no change
```
Comment (junrao) [inline core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala]: The comment seems incorrect. Topic deletion seems enabled and we are just enqueing a topic for deletion.

**Review by junrao — COMMENTED** [review]
@gitlw : Thanks for the updated patch. A few more comments.

File: core/src/main/scala/kafka/controller/TopicDeletionManager.scala (hunk 11 above)
Comment (gitlw) [reply core/src/main/scala/kafka/controller/TopicDeletionManager.scala]: Done.

File: core/src/main/scala/kafka/controller/TopicDeletionManager.scala (hunk 13 above)
Comment (gitlw) [reply core/src/main/scala/kafka/controller/TopicDeletionManager.scala:310]: I made this change in order to mock the method ReplicaStateMachine#handleStateChanges in the PartitionStateMachineTest.
Scala converts the handleStateChanges function into the following 2 java methods (reveald using javap ReplicaStateMachine.class):

public void handleStateChanges(scala.collection.Seq<kafka.controller.PartitionAndReplica>, kafka.controller.ReplicaState, kafka.controller.Callbacks);
public kafka.controller.Callbacks handleStateChanges$default$3();

and if the call here only passes in two parameters, the hidden method handleStateChanges$default$3() will be called for the 3rd parameter, and I haven't found a way to mock this hidden method in EasyMock.

File: core/src/main/scala/kafka/controller/TopicDeletionManager.scala (hunk 12 above)
Comment (gitlw) [reply core/src/main/scala/kafka/controller/TopicDeletionManager.scala]: (1) I changed the code to tracking ongoing deletions on the topic level.
(2) Good catch. I added the check to filter out topics to be deleted, as done in other methods including onReplicasBecomeOffline, triggerOnlinePartitionsStateChange etc.

File: core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala (hunk 14 above)
Comment (gitlw) [reply core/src/test/scala/unit/kafka/controller/PartitionStateMachineTest.scala]: Fixed.

File: core/src/main/scala/kafka/controller/KafkaController.scala [hunk 15]
```diff
@@ -1052,7 +1053,9 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
       debug(s"Live brokers: ${controllerContext.liveBrokerIds.mkString(",")}")
 
       val partitionsToActOn = controllerContext.partitionsOnBroker(id).filter { partition =>
-        controllerContext.partitionReplicaAssignment(partition).size > 1 && controllerContext.partitionLeadershipInfo.contains(partition)
+        controllerContext.partitionReplicaAssignment(partition).size > 1 &&
+          controllerContext.partitionLeadershipInfo.contains(partition) &&
```
Comment (junrao) [inline core/src/main/scala/kafka/controller/KafkaController.scala:1057]: Could we fix the alignment?

File: core/src/main/scala/kafka/controller/KafkaController.scala [hunk 16]
```diff
@@ -1052,7 +1053,9 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
...
+          !topicDeletionManager.isTopicQueuedUpForDeletion(partition.topic)
```
Comment (junrao) [inline core/src/main/scala/kafka/controller/KafkaController.scala:1058]: I think we also need to exclude the partitions being deleted in the replicatedPartitionsBrokerLeads(). Otherwise, the shutting down broker will keep retrying the controlled shutdown request on those partitions since their leaders will never be moved.

File: core/src/main/scala/kafka/controller/TopicDeletionManager.scala [hunk 17]
```diff
@@ -62,7 +62,27 @@ class TopicDeletionManager(controller: KafkaController,
...
+    * transition to or out of the OfflinePartition state.
+    *
+    * Hence we decide to change partition states to NonExistentPartition only when the actual deletion have started.
+    * For topics whose deletion have actually started, we keep track of them in the following topicsBeingDeleted
+    * variable. And once a topic is in the topicsBeingDeleted set, we are sure there will no longer
+    * be partition reassignments to any of its partitions, and only then it's safe to move its partitions to
+    * NonExistentPartition state. Once a topic is in the topicsBeingDeleted set, we will stop monitoring
+    * its partition state changes in the offlinePartitionCount metric
+    */
+  val topicsBeingDeleted = mutable.Set.empty[String]
```
Comment (junrao) [inline core/src/main/scala/kafka/controller/TopicDeletionManager.scala]: Is topicsBeingDeleted the same as topicsToBeDeleted -- topicsIneligibleForDeletion?

**Review by junrao — COMMENTED** [review]
@gitlw : Thanks for the new updates and the patience. Added a few more comments.

File: core/src/main/scala/kafka/controller/KafkaController.scala (hunk 15 above)
Comment (gitlw) [reply core/src/main/scala/kafka/controller/KafkaController.scala:1057]: Can you be more specific? If you mean the indentation by 2 spaces in the trailing conditions, it seems that's the correct alignment, as also being used in line 436 and line 1172 of this file.

Comment (junrao) [reply core/src/main/scala/kafka/controller/KafkaController.scala:1057]: Sorry. If there is the standard, we can just leave it as it is.

File: core/src/main/scala/kafka/controller/KafkaController.scala (hunk 16 above)
Comment (gitlw) [reply core/src/main/scala/kafka/controller/KafkaController.scala:1058]: For the partitions being deleted, the two behaviors are
1. If we exclude them in the response, then controlled shutdown will be allowed immediately even though the controller might need to issue StopReplica requests for deletion on the given broker, hence possibly blocking the topic deletion until the broker is brought up again.
2. If we instead include the partitions being deleted, the controller will block the controlled shutdown for a while (bound by controlled shutdown retries) and wait for topic deletion to finish. Once the topic deletion finishes, the controlled shutdown response will be empty. Or if the controlled shutdown retries are exhausted before the topic deletion can finish, the broker will be shutdown regardless of the response.
Hence I think the current behavior does not violate correctness. It's a matter of trade-off. Thoughts?

File: core/src/main/scala/kafka/controller/TopicDeletionManager.scala (hunk 17 above)
Comment (gitlw) [reply core/src/main/scala/kafka/controller/TopicDeletionManager.scala]: Not necessarily. While thinking about this problem, instead of topicsBeingDeleted, I think a better name is topicsWithDeletionStarted.

In the following sequence of events, the formula above does not always hold:
1. topic "A" enqueued for deletion
topicsToBeDeleted {A}
topicsWithDeletionStarted {}
topicsIneligibleForDeletion {}

2. starting topic deletion
topicsToBeDeleted {A}
topicsWithDeletionStarted {A}
topicsIneligibleForDeletion {}

3. While processing TopicDeletionStopReplicaResponseReceived, topic "A" is marked ineligible for deletion because the stopping replica failed
topicsToBeDeleted {A}
topicsWithDeletionStarted {A}
topicsIneligibleForDeletion {A}

4. A broker gets restarted, and it has replicas to be deleted:
topicsToBeDeleted {A}
topicsWithDeletionStarted {A}
topicsIneligibleForDeletion {}
...

File: core/src/main/scala/kafka/controller/KafkaController.scala (hunk 16 above)
Comment (junrao) [reply core/src/main/scala/kafka/controller/KafkaController.scala:1058]: Good analysis. This is my thought. Normally, topic deletion will happen very quickly. So, if we get to a state where the deletion of a topic is started, but not completed when a broker is being shut down, chances are that another broker is down, which prevents the topic deletion from completing. In this case, having the broker retry the controlled shutdown request won't be useful and only adds latency during shutdown.

File: core/src/main/scala/kafka/controller/TopicDeletionManager.scala (hunk 17 above)
Comment (junrao) [reply core/src/main/scala/kafka/controller/TopicDeletionManager.scala]: This makes sense.

File: core/src/main/scala/kafka/controller/KafkaController.scala (hunk 16 above)
Comment (gitlw) [reply core/src/main/scala/kafka/controller/KafkaController.scala:1058]: That makes sense. I've made the change. Thanks!

File: core/src/main/scala/kafka/controller/KafkaController.scala [hunk 18]
```diff
@@ -1076,7 +1079,9 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
         trace(s"All leaders = ${controllerContext.partitionLeadershipInfo.mkString(",")}")
         controllerContext.partitionLeadershipInfo.filter {
           case (topicPartition, leaderIsrAndControllerEpoch) =>
-            leaderIsrAndControllerEpoch.leaderAndIsr.leader == id && controllerContext.partitionReplicaAssignment(topicPartition).size > 1
+            !topicDeletionManager.isTopicQueuedUpForDeletion(topicPartition.topic) &&
```
Comment (junrao) [inline core/src/main/scala/kafka/controller/KafkaController.scala:1082]: It seems that we should be checking for isTopicWithDeletionStarted() instead of isTopicQueuedUpForDeletion()?

Comment (gitlw) [reply core/src/main/scala/kafka/controller/KafkaController.scala:1082]: At the detailed level:
Let's consider topics that are enqueued for deletion but haven't started the actual deletion, maybe they have ongoing partition reassignments that would still take many hours to finish. And for those partitions during a controlled shutdown we won't try to switch leaderships because we no longer care about their availability.
And here even if the shutting down broker is still leader for those partitions, I think it's ok to shutdown the broker immediately because again we don't care about the partition availability anymore. Using !isTopicWithDeletionStarted will still return those partitions and block the controlled shutdown, which is probably not ideal.

At a higher level:
isTopicQueuedUpForDeletion is used in many other places so that for other types of operations, once a topic is enqueued, it effectively no longer exists. I think that also applies here, and using the same method consistently simplifies reasoning. Thoughts?

Comment (junrao) [reply core/src/main/scala/kafka/controller/KafkaController.scala:1082]: Thanks for the explanation. This sounds good then.

**Review by junrao — APPROVED** [review]
@gitlw Thanks for the updated patch. LGTM
